File: Java/202. Happy Number.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();
        
        while(!seen.contains(n)){
            seen.add(n);
            int sum = 0;
            
            /* calculate the sum of the squares of each digit
             * using modulo instead of reading strings
             */
            while(n > 0){
                sum += (n%10)*(n%10);
                n /= 10;
            }
            
            if(sum == 1){
                return true;
            }
            
            n = sum;
        }
        
        return false;
    }
}

File: Python3/202. Happy Number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()
        while n not in seen:
            # add it to the set
            seen.add(n)
            
            # refresh the sum
            s = 0
            
            # calculate the sum of the squares of the digits
            for c in str(n):
                s += int(c)*int(c)
                
            # check if s equals to 1
            if s == 1:
                return True
            
            # update the value of n
            n = s
            
        # the loop ends because n is in seen
        return False